### ISSUE TITLE
Feature request: Enhanced Backpacks

### ISSUE THREAD

**Issue by 63502** [copygirl/WearableBackpacks#2]
Firstly, I am really glad that someone finally ported the backpacks that better storage added. They are in my opinion the most exceptionally well made and "realistic" backpacks added by any mod, and it was a huge disappointment to see them left behind in 1.7.10 (not anymore of course XD). I am thrilled to see them return, and I hope you will be open to bringing back more than just the standard backpacks.

In the better storage mod, there were two special backpacks as well as the normal ones.

1. The first type is called the Ender Backpack. Like its name suggests, it is an ender chest that the player can carry as a backpack. There is no crafting recipe in survival, so there is only one way to get one. A player that wants one must find an enderman in the world that is wearing one, place an ender chest nearby, and get that enderman to pick up the ender chest. If the enderman picks up the chest, the backpack will be placed down where the chest was. The player can then equip the backpack by breaking it while holding shift like any other backpack. If the player attempts to break the pack without holding shift, the pack will teleport away like a dragon egg.
Also if I recall correctly, The endermen that wear ender backpacks are a separate entity added by better storage. They are called Friendly Endermen and they do not attack the player and can never be provoked.

2. The second backpack is actually part of the better storage mod's integration with Thaumcraft 4. The Thaumaturge's backpack is a backpack with 2 or 3 times more slots than the normal backpack. Obviously having the thaumaturge's backpack is not possible at this point in time since TC6 is still in development. With that said though, it would be nice to have some larger mod integrated backpacks that require special crafting methods. A few ideas would be to make a bound backpack for Blood Magic and an end game Gaia backpack for Botania. Perhaps when the time finally rolls around, the Thaumaturge's Backpack can make an appearance when TC6 gets released too.

One last side note, the original Better storage had a config option to enable/disable a keybind that allows the player to open a worn backpack without taking it off. I believe it was disabled by default.

I hope you like what you see. :)

Thanks,
     63502

**Comment by copygirl** [comment]
I'm currently not actively working on the mod but these are certainly features I'd like to re-implement.

- **Ender Backpacks** - Before I can tackle these, I have to finish the part of the mod that allows non-player entities to wear backpacks. While the bulk of the work is done. The friendly enderman will make a return, though I'm not sure if I'm keeping the obscure and annoying trading mechanic.
- **Thaumaturge's Backpack** - The texture is already (or rather, still) there. What I hope I can do is implement a backpack for every major mod, each with its own mod-related feature. It works well to reward the player for getting into these mods and help a little with the large amount of unique items when playing with lots of mods. The size of those backpacks was just 44% larger than the normal one, by the way, though this time around I plan to make the sizes completely adjustable (per item, and both width and height). I also plan to add specialized backpacks that don't hold items, like a fluid backpack? (Might be modeled as a barrel?)
- **Key to open equipped backpack** - Technically easily possible, though I want to implement config setting synchronization when joining a server.

**Comment by laserlemons** [comment]
For what it's worth, I think the enderman trading mechanic was fun and unique. Much better than just crafting it, it made the ender backpack something to be proud of.

**Comment by 63502** [comment]
Indeed. I only succeeded at doing it in survival once.

**Comment by copygirl** [comment]
Small update: Opening backpacks using a key has been implemented in 6e613203f640d70a0b951d5537bb4edf4aebc57e.

**Comment by TDarkShadow** [comment]
Maybe a kinda stupid suggestion but what about tiered / upgradeable backpacks?
Like a Iron / Golden / Diamond Backpack?

Otherwise something along the lines that has the same armor points as an Iron / Diamond chestplate?
Simply Jetpacks solved this by adding armor plating so you could have the same leather backpack while having increased armor points.

**Comment by copygirl** [comment]
> Maybe a kinda stupid suggestion but what about tiered / upgradeable backpacks?
Like a Iron / Golden / Diamond Backpack?

The only tiered backpacks I'm going to add are ones gated behind other mods' progression trees, ones where you actually have to put in effort and time to create, instead just relatively easy to get Vanilla resources. Otherwise... you're just going to make the highest tier backpack and where's the fun in that? I **do** advertise my mod for being balanced, after all.

Also, ideally, they'd serve some function together with the other mod, like allowing certain items from it to still function or draw resources from the backpack. A Thaumcraft backpack might allow items to repair inside it, and also be compatible with the Repair enchantment itself. A Botania backpack would allow you to draw mana from items inside the backpack, and repair automatically using mana, as well. Stuff along those lines. Maybe I can come up with some neater, unique features for them, too.

Alternatively, people can use MineTweaker or similar to add a recipe for a [backpack with a custom size](https://twitter.com/koppeh/status/832019349460361216).

> Otherwise something along the lines that has the same armor points as an Iron / Diamond chestplate?

Again, if you make it an option, people will just use it, and that destroys a big balance point in the mod: Trading armor for storage. Also, with the config option that allows you to wear both armor and a backpack, that feature is already in the mod in spirit, better than I could ever implement "armor plating": You can wear **any** armor you like at the same time.

----

Now, with both of those suggestions separately out of the way, I **had** considered adding an armored backpack with more defense but less storage, though I decided against it since it simply doesn't feel right.

Ender backpacks will likely end up being the only other item storage backpack you'll be able with just Vanilla, unless I come up with a unique feature for another one. Liiike... one that's only able to store shulker boxes, but allows easy access to them without having to place them?

Other than that, I'm planning to use enchantments (likely dungeon loot?) to add additional features, instead of relying on new backpack types or upgrades in the form of items or crafting recipes.

**Comment by TDarkShadow** [comment]
You're right and I appreciate your well-constructed feedback and opinion.
I entirely agree with your balancing mechanics and I'm looking forward for future additions.

>A Botania backpack would allow you to draw mana from items inside the backpack.

While this sounds very handy, I think Botania already adds something in that line?
http://botaniamod.net/lexicon.php#baubleBox
Not sure if it lets you draw mana from it but I believe it does.

>I had considered adding an armoured backpack with more defense but less storage / unless I come up with a unique feature for another one.

I would argue, rather than one to store shulker boxes, to maybe add a way (as an anvil-enchant or otherwise) to add an Elytra to your Backpack but removes the enchantability from it.
Aside from armor, a portable 3x3 crafting grid is always nice to have but maybe that is getting the mod a little bloated for what promotes.

**Comment by copygirl** [comment]
> I would argue, rather than one to store shulker boxes, to maybe add a way (as an anvil-enchant or otherwise) to add an Elytra to your Backpack but removes the enchantability from it.

I don't think anyone would use an elytra backpack in Vanilla if it's just going to break easily. Also that's the same thing as with the armor plating in the end. Mods doing this I find are just taking a very easy, non-dynamic approach.

If in Vanilla end-game, you have to decide between lots of storage (shulker boxes add to that - storing them in your backpack will allow massive amounts of storage) and the ability to fly, that's exactly the kind of hard decision I want players to make. Storage or armor. Flying or massive storage. Players who don't necessarily care about the design decision behind this have the choice to change the config setting allowing you to wear both backpacks and armor.

I simply don't think there's any better way to implement either a balanced or a "beautiful" way to mush these together, let alone both. Beautiful as in.. not sure how to explain, but avoiding hardcoded approaches like X + Y = XY, simply crafting and upgrades without real drawback.

The only thing I could imagine is an upgrade or enchantment that would be beautiful, but not necessarily balanced, is a way to upgrade the backpack so it doesn't take up your chest armor slot anymore (just as if the config setting was off). But that's not really balanced anymore. Maybe I'll add an NBT setting that would do the same thing, and add per-backpack options? This way, modpack creators could fiddle with the balance themselves, too.

And no, I'm not the kind of person who gates (in my opinion) overpowered things behind end-game content such as nether stars. I prefer end-game unlocking new, fun ways of playing the game without invalidating the old.

Elytra + fireworks for example I don't like, no matter how fun they are. Who needs to walk *anywhere* if you can just fly? Horses? Minecarts? Pff! Without the instant boost, people came up with neat launchers and ways to get around, now flying is just a *rightclick* away! Blegh. `</rant>`

> Aside from armor, a portable 3x3 crafting grid is always nice to have but maybe that is getting the mod a little bloated for what promotes.

Oh yes, that's true. An item storage backpack that comes with a crafting grid is certainly worth considering.

**Comment by PeakXing** [comment]
1 hope that the arrow in the backpack can be fired.
2 would like to join the bag tied to the leg.

**Comment by copygirl** [comment]
> hope that the arrow in the backpack can be fired.

A backpack is not a quiver and I don't know if appropriate hooks exist to override bow drawing behavior.

> would like to join the bag tied to the leg.

If I understand correctly... A belt pouch would be neat but it both is difficult to design a natural way to interact with it and it sort of doesn't fit into the mod. Also the uses I could come up with for it can already be found in other mods, such as a belt that allows you to switch your hotbar with the belt's inventory.

**Comment by Xorbah** [comment]
So is this Ender backpack still planned? Would be a phenomenal addition and would for all intents are purposes perfect this mod in standard modpacks.

**Comment by copygirl** [comment]
This would be up to the current maintainers to decide, if they can get the time and motivation. I don't quite know anymore how much code is in place to support multiple types of backpacks easily.
